### ISSUE TITLE
Reference Target: How should we treat invalid reference targets for relations set via Element IDL attributes?

### ISSUE THREAD

**Issue by alice** [WICG/webcomponents#1089]
#1071 discusses how invalid reference targets should be handled when computing the browser-internal ("true") target of an ID-based relationship. The consensus in that issue is that an invalid reference target:
- should be treated browser-internally like an invalid ID, i.e. as if the attribute value was itself an invalid ID; and
- should be treated likewise by read-only IDL attribute getters like `label.control`; i.e. those getters should return null even though an element with the ID in question technically exists in the same tree.

How should attributes _set_ via IDL attributes work in this situation?

1) I think you could make an argument that you should _always_ get back the element that you set, assuming that element is a descendant of a shadow-including ancestor.
2) However, I think you could also make an argument that it should only return the element that you set if it is also a valid value - i.e. that it has no referenceTarget set, or has a valid referenceTarget set. This would be analogous to how you can set an IDL attribute to be an element that is not a descendant of a shadow-including ancestor, but it won't be returned from the getter.

My personal slight preference would be for (2), though I could easily be persuaded otherwise. This would make those IDL attribute getters marginally slower, as we would need to check validity before returning the appropriate value. However, it would also give an indication that the relationship has not been set successfully, and work analogously to the read-only IDL attributes.

---

To address the original question in this issue, there seems to be a consensus based on the discussion on the earlier text of this issue, and also the [WPT tests](https://github.com/web-platform-tests/wpt/blob/79bdb1b12acd159e80b2186b3180d459734ba999/shadow-dom/reference-target/tentative/aria-labelledby.html#L56), that:
- relationships set up via IDL attribute setters like: 
   ```js
   el.ariaLabelledByElements = [ customElementWithReferenceTarget ];
   ```
   should work identically to relationships set via content attributes like: 
   ```html
   <input aria-labelledby="id-for-custom-element">
   ```
   i.e. that the referenceTarget should be followed in both cases; and
- when the reference target is _valid_, that the IDL attribute getters should return the element[s] that were set:
   ```js
   const labels = [ customElementWithReferenceTarget ];
   el.ariaLabelledByElements = labels;
   el.ariaLabelledByElements[0] === labels[0];
   ```
   i.e. that the getters should not return elements inside shadow roots, even if those elements are the "true" target of the relationship.

<details>
<summary>Original text</summary>

The explainer [proposes](https://github.com/WICG/webcomponents/blob/gh-pages/proposals/reference-target-explainer.md#javascript-attributes-that-reflect-element-objects) that IDL attribute _getters_ not return the `referenceTarget` element, and I strongly agree there.

However, this leaves open the question of what should happen with IDL attribute _setters_.

```html
<input id="input">
<fancy-listbox id="listbox">
  <template
    shadowrootmode="open"
    shadowrootreferencetarget="real-listbox"
  >
    <div id="real-listbox" role="listbox">
      <div id="option-1" role="option">Option 1</div>
      <div id="option-2" role="option">Option 2</div>
    </div>
  </template>
</fancy-listbox>
```

```js
input.ariaControlsElements = [ listbox ];
```

In this case, should the _internal_ `aria-controls` relationship (i.e. what is exposed to the accessibility tree) be with the `<fancy-listbox>`, or the `<div role="listbox">`?

</details>

**Comment by alice** [comment]
I see the WPT tests assert that IDL attributes _do_ follow `referenceTarget` references, e.g.:

https://github.com/web-platform-tests/wpt/blob/master/shadow-dom/reference-target/tentative/anchor.html#L105
https://github.com/web-platform-tests/wpt/blob/79bdb1b12acd159e80b2186b3180d459734ba999/shadow-dom/reference-target/tentative/aria-labelledby.html#L56

**Comment by alice** [comment]
Another question related to this and to #1071: what should happen in this case?

```html
<input id="input">
<fancy-listbox id="listbox">
  <template
    shadowrootmode="open"
    shadowrootreferencetarget="BAD_ID"
  >
    <div id="real-listbox" role="listbox">
      <div id="option-1" role="option">Option 1</div>
      <div id="option-2" role="option">Option 2</div>
    </div>
  </template>
</fancy-listbox>
```

```
input.ariaControlsElements = [ listbox ];
console.log(input.ariaControlsElements);  // what does this log?
```

Should it log the `fancy-listbox`, even though the reference target is invalid? Or should it log an empty list, _because_ the reference target is invalid?

**Comment by annevk** [comment]
I think APIs should generally return the custom element (that handles the forwarding into its shadow tree). It seems weird if the APIs didn't work just because you used a custom element (with forwarding).

For the later question I think we have to answer another question. When do we consider a custom element a viable ID target?

1. Always.
2. Only when it has a reference target defined.
3. Only when it has a valid reference target defined.

It seems potentially somewhat performance-sensitive to do 3, but not sure.

**Comment by annevk** [comment]
This might also duplicate #1071?

**Comment by alice** [comment]
I'm not quite sure I follow your answers, sorry @annevk.

The question I was trying to ask in the issue description, which I didn't explain very well in hindsight, was whether relationships set via Element IDL attributes (as opposed to content attributes which necessarily use ID strings) should also follow reference target forwarding when computing the relationships. 

The explainer doesn't address this; it only says:

> This feature is intended to work with all attributes that refer to another element by ID string.

and

> These [IDL attribute getters] will _never_ directly return the referenceTarget element that's inside the shadow tree. 

The explainer is clear on how the IDL attribute getters should work _when the referenceTarget is valid_: they should return the element in the appropriate scope (i.e. the element which was set via the setter, which must be a descendant of a shadow-including ancestor of the element hosting the attribute).

While the explainer doesn't address how the IDL attribute setters should work, the WPTs indicate that the intent was for attribute set via IDL attributes to work similarly to attribute set via content attributes.

> I think APIs should generally return the custom element (that handles the forwarding into its shadow tree). It seems weird if the APIs didn't work just because you used a custom element (with forwarding).

When you say the APIs should return the custom element, are you answering the later question in https://github.com/WICG/webcomponents/issues/1089#issuecomment-2533164216, regarding whether the IDL attribute getters should return null/empty list for an invalid referenceTarget?

> This might also duplicate https://github.com/WICG/webcomponents/issues/1071?

The reason I think that this is a slightly different question to #1071 is that #1071 (part 2) uses a getter-only API as an example. #1071 seems to conclude that an invalid referenceTarget is analogous to a bad ID, so even if the ID reference to the host is valid, the attribute is invalid. Because the attribute is invalid, the IDL getter returns null, just as it would if the content attribute was a bad ID.

With an IDL attribute setter, there is no ID involved, so I think you could make an argument that you should always get back the element that you set, assuming that element is a descendant of a shadow-including ancestor.

However, I think you could also make an argument that it should only return the element that you set if it is also a _valid_ value, per my reasoning below. This would be analogous to how you can _set_ an IDL attribute to be an element that is not a descendant of a shadow-including ancestor, but it won't be returned from the getter.

> For the later question I think we have to answer another question. When do we consider a custom element a viable ID target?

I think a custom element can be a viable ID target when either:

1) It is, itself, a viable ID target, and doesn't have a referenceTarget set. 
   For example:
   ```html
   <button popovertarget="custom">Show popover</button>
   <custom-popover id="custom">
     <template shadowrootmode="open">
       <!-- no referenceTarget -->
       Some popover content.
     </template>
   </custom-popover>
   ```

2) Or, it contains a viable ID target, and has a valid referenceTarget set referring to the viable target.
   For example:
   ```html
   <label for="custom">Given name</label>
   <custom-input id="custom">
     <template shadowrootmode="open" shadowrootreferencetarget="real-input">
       <input id="real-input">
     </template>
   </custom-input>
   ```
   
   Without the referenceTarget forwarding to the `<input>`, `custom` is not a valid target for `for`, as it is not a labelable element. It would be equally invalid whether the referenceTarget was missing, or was a bad ID, or referred to an element which was not a labelable element.

> It seems potentially somewhat performance-sensitive to do 3, but not sure.

Depending on the implementation, this could be performance sensitive when resolving an ID target (note that this is different from either `getElementById()` or a query selector), or require extra storage to keep track of both the "shallow" and the "deep" version of each ID reference, plus update them on relevant mutations. However, I think in practice the performance difference should be negligible. It would at worst require a small number of extra lookups (two per shadow root - one to check the presence of a referenceTarget, and one to resolve it) when using certain DOM APIs.

Regardless of the decision we make for DOM APIs, we will still need to implement that validity logic for whatever the consequences are of the relation. For example, a `for` attribute on a `<label>` only has any effect when it refers to a labelable element. When determining the activation behaviour of the `<label>`, we will still need to determine what "deep" element it refers to.

**Comment by dandclark** [comment]
I'm still a little uncertain on what the invalid ID behavior should look like, but assuming valid IDs I think attribute *setters* should cause `referenceTarget` associations to be followed. I don't see any reason not to support this, and it seems like it'd be confusing for developers if `referenceTarget` only worked when the element relationships were established via attribute.

**Comment by alice** [comment]
@dandclark Yeah, I agree. I've updated the issue to focus more on the invalid referenceTarget question, and to make note of the consensus on the earlier question.

**Comment by dandclark** [comment]
Given the tentative conclusion that #1071 seems to be converging towards (treating the reference as totally invalid rather than referring to the shadow host), I agree @alice that your option (2) is a better fit:

> 2. However, I think you could also make an argument that it should only return the element that you set if it is also a valid value - i.e. that it has no referenceTarget set, or has a valid referenceTarget set. This would be analogous to how you can set an IDL attribute to be an element that is not a descendant of a shadow-including ancestor, but it won't be returned from the getter.

If I’m understanding correctly, that means we’d get behavior like this:

```html
<input id="input">
<fancy-listbox id="listbox">
  <template
    shadowrootmode="open"
    shadowrootreferencetarget="BAD_ID"
  >
    <div id="real-listbox" role="listbox">
      <div id="option-1" role="option">Option 1</div>
      <div id="option-2" role="option">Option 2</div>
    </div>
  </template>
</fancy-listbox>
<script>
input.ariaControlsElements = [ listbox ];
console.log(input.ariaControlsElements);  // logs `null`
document.querySelector("#fancy-listbox").shadowRoot.referenceTarget = "real-listbox";
console.log(input.ariaControlsElements);  // logs `listbox` since the reference target is valid now!
document.querySelector("#fancy-listbox").shadowRoot.referenceTarget = "";
console.log(input.ariaControlsElements);  // logs `listbox` since there is no reference target; aria-controls now applies to the shadow host
</script>
```

Which seems reasonable given the validity checks that already exist in element-reflecting IDL attributes.

**Comment by alice** [comment]
@dandclark Yes to all of that, except that I'm a little unclear on this line: 

```js
document.querySelector("#fancy-listbox").shadowRoot.referenceTarget = "";
```

I would have expected an empty string to work like an invalid ID, rather than like an un-set. To un-set, I would have expected we'd need to set `referenceTarget` to `null`, similar to setting an attribute to `null` to remove it.

**Comment by dandclark** [comment]
@alice To make sure I understand you, are you thinking that the `referenceTarget` attribute's type should be a nullable `DOMString`?

```js
partial interface ShadowRoot {
  attribute DOMString? referenceTarget;
}
```

I had been assuming the attribute's type would just be `DOMString`, and that's how it's [prototyped today](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/dom/shadow_root.idl;l=58;drc=86547d0ef65244867832bba1721ef3d0ed826eea). But looking at the explainer again I guess it's not specific on this point.

Assuming I'm reading the specs correctly, keeping it as a non-nullable `DOMString` triggers [this string conversion](https://webidl.spec.whatwg.org/#js-DOMString) when trying to set the attribute to null, and this conversion happens before the implementation of the property "sees" the value:
```js
document.querySelector("#fancy-listbox").shadowRoot.referenceTarget = null;
document.querySelector("#fancy-listbox").shadowRoot.referenceTarget; // logs 'null'
```
But if we make it `DOMString?` then we'd get the behavior you suggest. That's definitely seems like it'd avoid a pitfall when someone wants to clear the referenceTarget value. An inconsistency it introduces though is that the `Element.id` property is itself a non-nullable `DOMString`. So you get this today:

```js
const div = document.createElement("div");
div.id = null;
div.id; // logs 'null'
```

It's tempting to treat the `referenceTarget` property as kind of equivalent to setting an ID elsewhere, so keeping the behavior consistent could be nice. What do you think? (Maybe this question should be its own thread.)

**Comment by alice** [comment]
@dandclark I think that is indeed worthy of a separate thread :)

**Comment by dandclark** [comment]
Added WPT coverage for this here: https://github.com/web-platform-tests/wpt/pull/52295. @alice please let me know if you disagree with any of the behavior suggested in those tests (aside from the question of the type forked into #1093). That WPT is now basically testing for the behavior suggested in https://github.com/WICG/webcomponents/issues/1089#issuecomment-2594204895.
